### PR TITLE
`merkledb` -- limit number of goroutines calculating node IDs

### PR DESCRIPTION
**Pull request by danlaine**
## Why this should be merged

Resolves TODO. Alternate to #1957.

## How this works

Uses a semaphore to control the number of goroutines calculating node IDs (namely the number of goroutines executing `calculateNodeIDsHelper`) across all views. 

In `calculateNodeIDs`, we will block until we acquire from the semaphore before calling `calculateNodeIDsHelper`.
In `calculateNodeIDsHelper`, we call `calculateNodeIDsHelper` in a goroutine iff the semaphore has unused resources. Otherwise, we call `calculateNodeIDsHelper` recursively.

## How this was tested

Existing UT

### CODE REVIEW

File: x/merkledb/trieview.go [hunk 1]
```diff
@@ -259,18 +257,17 @@ func (t *trieView) calculateNodeIDs(ctx context.Context) error {
...
 	var (
 		// We use [wg] to wait until all descendants of [n] have been updated.
-		// Note we can't wait on [eg] because [eg] may have started goroutines
-		// that aren't calculating IDs for descendants of [n].
 		wg              sync.WaitGroup
 		updatedChildren = make(chan *node, len(n.children))
 	)
 
 	for childIndex, child := range n.children {
-		childIndex, child := childIndex, child
```
Comment (danlaine) [inline x/merkledb/trieview.go:272]: Don't think this was actually necessary

**Review by darioush — APPROVED** [review]

File: x/merkledb/trieview.go [hunk 2]
```diff
@@ -279,22 +276,22 @@ func (t *trieView) calculateNodeIDsHelper(ctx context.Context, n *node, eg *errg
...
 		}
 
 		// Try updating the child and its descendants in a goroutine.
-		if ok := eg.TryGo(func() error {
-			updateChild()
-			return nil
-		}); !ok {
+		if ok := t.db.calculateNodeIDsSema.TryAcquire(1); ok {
+			go calculateChildID()
+			t.db.calculateNodeIDsSema.Release(1)
```
Comment (dboehm-avalabs) [inline x/merkledb/trieview.go:291]: shouldn't this be part of the goroutine and deferred?

File: x/merkledb/trieview.go [hunk 3]
```diff
@@ -231,21 +231,19 @@ func (t *trieView) calculateNodeIDs(ctx context.Context) error {
...
 		}
 
-		// [eg] limits the number of goroutines we start.
-		var eg errgroup.Group
-		eg.SetLimit(t.db.rootGenConcurrency)
-		t.calculateNodeIDsHelper(ctx, t.root, &eg)
-		if err = eg.Wait(); err != nil {
-			return
-		}
+		_ = t.db.calculateNodeIDsSema.Acquire(context.Background(), 1)
```
[the remark this answers is not in the document]
Comment (danlaine) [reply x/merkledb/trieview.go:244]: Not sure I understand the question. In order for a new goroutine to run `calculateNodeIDsHelper`, it must acquire 1 from the semaphore and release that 1 when it exits the goroutine.

File: x/merkledb/trieview.go (hunk 2 above)
Comment (danlaine) [reply x/merkledb/trieview.go:291]: fixed

**Review by dboehm-avalabs — APPROVED** [review]

File: x/merkledb/db.go [hunk 4]
```diff
@@ -125,8 +126,8 @@ type Config struct {
 	// RootGenConcurrency is the number of goroutines to use when
 	// generating a new state root.
 	//
-	// If 0 is specified, [runtime.NumCPU] will be used. If -1 is specified,
-	// no limit will be used.
+	// If 0 is specified, [runtime.NumCPU] will be used.
+	// Must be non-negative.
 	RootGenConcurrency int
```
Comment (dboehm-avalabs) [inline x/merkledb/db.go:131]: should make uint if must be positive

Comment (danlaine) [reply x/merkledb/db.go:131]: Yeah I noticed the same. There are a bunch of other config fields that should uints. I will make a follow up PR to change them.

Comment (danlaine) [reply x/merkledb/db.go:131]: Ill change this to a uint though

**Pull request merged**